tests/lean/run/6927.lean

```diff
@@ -0,0 +1,115 @@
+import Lean
+/-
+# Nested `let rec` declarations in the presence of delayed metavariable assignments
+
+https://github.com/leanprover/lean4/issues/6927
+
+Nested declarations using `let rec` should compile correctly even when nested within expressions
+that are elaborated using delayed metavariable assignments, such as `match` expressions and tactic
+blocks. Previously, declaring a `let rec` within such an expression in the following fashion
+```lean
+def f x :=
+  let rec g :=
+    match ... with
+    | pat =>
+      let rec h := ... g ...
+      ... x ...
+```
+where `g` depends on some free variable bound by `f` (like `x` above) would cause `MutualClosure` to
+fail to detect that transitive fvar dependency of `h` (which must pass it as an argument to `g`),
+leading to an unbound fvar in the body of `h` that was ultimately fed to the kernel. This occurred
+because `MutualClosure` processes let-recs from most to least nested. Initially, the body of `g` is
+an application of the delayed-assigned metavariable generated by `match` elaboration; the root
+metavariable of that delayed assignment is, in turn, assigned to an expression that refers to the
+mvar that will eventually be assigned to `g` once we process that declaration but for now is
+unassigned. Therefore, when we initially process the most-nested declaration `h` (before `g`), we
+cannot instantiate the `match`-expression mvar's delayed assignment (since the metavariable that
+will eventually be assigned to the yet-unprocessed `g` remains unassigned). Thus, we do not detect
+any of the fvar dependencies of `g` in the `match` body -- namely, that corresponding to `x`, which
+`h` should therefore also take as a parameter.
```

Comment:
Good comment! Maybe better to put that as a `[Note]` (GHC-style) at the top or bottom of `Lean/Elab/MutualDef.lean` and link it from the the relevant code line.

(See “Documentation & GHC Notes” in our onboarding manual)